packages/nextjs-contentful-renderer/src/getPageContent/getPageContent.ts

```diff
@@ -19,18 +20,21 @@ export const getPageContent = async (
 
   const existingPageId = getVisitedPageIdFromCookies(context, slug);
   if (existingPageId) {
-    const page = await getPageById(apolloClient, existingPageId, preview);
+    const { modelData, pageId } = existingPageId;
+    const page = await getPageById(apolloClient, pageId, preview, modelData);
     if (page) {
       return page;
     }
   }
 
-  const route = await getRouteBySlug(apolloClient, slug, preview, domain);
-  if (!route) return undefined;
+  const replicaRoute = await getRouteBySlug(apolloClient, slug, preview, domain);
+  if (!replicaRoute) return undefined;
 
-  const pageId = getPageId(route, context.query.utm_campaign);
-  const page = await getPageById(apolloClient, pageId, preview);
-  setCookie(context, slug, pageId);
+  const modelData = Object.keys(replicaRoute).includes('modelData') ? (replicaRoute as Replica).modelData : null;
+
+  const pageId = getPageId(replicaRoute, context.query.utm_campaign);
+  const page = await getPageById(apolloClient, pageId, preview, modelData);
+  setCookie(context, slug, { modelData, pageId });
```

Comment:
what's included in the `modelData`?, AFAIK cookies have a max length, this can _easily break_. Also, are we 100% this is not leaking any private data?

Reply:
Yeah, we might need to rethink this one then. The `modelData` would contain the entire JSON for the `blocks`/`models` within a `Replica`/`Page`. I don’t foresee any private data being leaked, since it is the content that would make it to the page, but it could easily exceed the max length.